### PR TITLE
Wait to save thumbnail before close the project

### LINKED ISSUE #7874: Project thumbnail isn't generated upon closing the project and saving
### Steps to reproduce

1. Open a project
2. Make some changes to it (e.g. trim a clip)
3. Ctrl + W
4. Choose "Save"
5. Review the thumbnail of the project

### Expected behavior

The thumbnail is updated and reflects the current state of the project.

### Actual behavior

The thumbnail isn't updated

### Audacity Version

Audacity 4 alpha

### Operating system

Windows, macOS, Linux

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by embarc-gabriel**
Resolves: #7874

A callback is called once the thumbnail is ready to be saved on the disk.
Once this actions takes some time, it was being executed after navigation was performed and the project closed.

The current solution implement the thumbnailCreator class to address this issue.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/project/iaudacityproject.h [hunk 1]
```diff
@@ -39,7 +39,8 @@ class IAudacityProject
     virtual void setNeedAutoSave(bool val) = 0;
     virtual muse::async::Notification needSaveChanged() const { return muse::async::Notification(); }
     virtual muse::Ret save(const muse::io::path_t& path = muse::io::path_t(), SaveMode saveMode = SaveMode::Save) = 0;
-    virtual muse::async::Notification captureThumbnailRequested() const = 0;
+    virtual muse::async::Channel<std::string> captureThumbnailRequested() const = 0;
+    virtual void onThumbnailCreated(bool success) = 0;
```
Comment (igorkorsukov) [inline src/project/iaudacityproject.h:43]: I suggest removing these methods from the project. In ProjectPropertiesModel, use the creator directly.
There is such logic here, the public interface for creating a thumb is the IThumbnailCreator. The ProjectPropertiesModel model is its implementation...
In the future, we can change the implementation, for example, to create Thumbnail, use not the current Qml scene grab, but independent drawing or something like that.

File: src/project/internal/audacityproject.cpp [hunk 2]
```diff
@@ -266,20 +272,28 @@ Ret Audacity4Project::doSave(const muse::io::path_t& savePath, bool generateBack
...
+    if (!ret) {
+        return make_ret(Ret::Code::UnknownError);
     }
-    return muse::make_ret(muse::Ret::Code::UnknownError);
+
+    if (createThumbnail) {
+        std::string thumbnailPath = savePath.toStdString();
+        thumbnailPath.replace(thumbnailPath.size() - 4, 4, "png");
+
+        Ret ret = thumbnailCreator()->createThumbnail(thumbnailPath);
```
Comment (igorkorsukov) [inline src/project/internal/audacityproject.cpp:289]: There are two things that bother me here: 
1. What if a path without an extension is passed here, or with some other extension, something like `file.au.backup`... I think it's better to use `muse::io::FileInfo` for path parsing 
2. the fact that the thumbnail is saved in PNG format is something only the creator “knows”, i.e. we either reveal the knowledge here, or we tell the creator to save it in PNG, and not in any other way.

File: src/project/internal/thumbnailcreator.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include "thumbnailcreator.h"
+
+using namespace au::project;
+
+void ThumbnailCreator::onThumbnailCreated(bool success)
+{
+    m_thumbnailCreated.send(success);
+}
+
+muse::async::Channel<std::string> ThumbnailCreator::captureThumbnailRequested() const
```
Comment (igorkorsukov) [inline src/project/internal/thumbnailcreator.cpp:10]: better to use `muse::io::path_t` to indicate what is file path

File: src/project/internal/thumbnailcreator.cpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+{
+    return m_createThumbnailRequested;
+}
+
+muse::Ret ThumbnailCreator::createThumbnail(const std::string& path)
+{
+    muse::Ret ret;
+    QEventLoop loop;
+    m_thumbnailCreated.resetOnReceive(this);
```
Comment (igorkorsukov) [inline src/project/internal/thumbnailcreator.cpp:19]: I think it's better to move it after `loop.exec()`.
There may be a potential problem here, in `onReceive` we pass a lambda in which we save a reference to a local  variable (loop). If someone for some reason calls `onThumbnailCreated`, and `onReceive` is triggered, then there may be a crash. Therefore, it is safer for us to simply delete this callback after it is triggered.

File: src/projectscene/qml/Audacity/ProjectScene/clipsview/TracksClipsView.qml [hunk 5]
```diff
@@ -29,12 +29,13 @@ Rectangle {
     ProjectPropertiesModel {
         id: project
 
-        onCaptureThumbnail: function captureThumbnail() {
+        onCaptureThumbnail: function captureThumbnail(thumbnailUrl) {
             // hide playCursor for the time grabbing image
             playCursor.visible = false
             content.grabToImage(function(result) {
                 playCursor.visible = true
-                result.saveToFile(project.thumbnailUrl)
```
Comment (igorkorsukov) [inline src/projectscene/qml/Audacity/ProjectScene/clipsview/TracksClipsView.qml:37]: better to remove `thumbnailUrl` property from ProjectPropertiesModel

Comment (embarc-gabriel) [reply src/projectscene/qml/Audacity/ProjectScene/clipsview/TracksClipsView.qml:37]: Done.

File: src/project/internal/thumbnailcreator.cpp (hunk 4 above)
Comment (embarc-gabriel) [reply src/project/internal/thumbnailcreator.cpp]: Good catch. Done.

File: src/project/internal/thumbnailcreator.cpp (hunk 3 above)
Comment (embarc-gabriel) [reply src/project/internal/thumbnailcreator.cpp]: Done.

File: src/project/internal/audacityproject.cpp (hunk 2 above)
Comment (embarc-gabriel) [reply src/project/internal/audacityproject.cpp]: Ok. I have created a helper method and it is way better now.

File: src/project/iaudacityproject.h (hunk 1 above)
Comment (embarc-gabriel) [reply src/project/iaudacityproject.h]: Yes! I have refactor it.

File: src/project/internal/thumbnailcreator.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+class ThumbnailCreator : public IThumbnailCreator, public muse::async::Asyncable
+{
+public:
+    ThumbnailCreator() = default;
+    void onThumbnailCreated(bool success) override;
+    muse::Ret createThumbnail(const muse::io::path_t& path) override;
+    muse::async::Channel<muse::io::path_t> captureThumbnailRequested() const override;
+
+private:
+    muse::io::path_t thumbnailPath(const muse::io::path_t& path);
```
Comment (igorkorsukov) [inline src/project/internal/thumbnailcreator.h:19]: it can be marked as `const` method

File: src/project/ithumbnailcreator.h [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+class IThumbnailCreator : MODULE_EXPORT_INTERFACE
+{
+    INTERFACE_ID(au::project::ithumbnailcreator)
+
+public:
+    virtual ~IThumbnailCreator() = default;
+    virtual void onThumbnailCreated(bool success) = 0;
+    virtual muse::Ret createThumbnail(const muse::io::path_t& path) = 0;
+    virtual muse::async::Channel<muse::io::path_t> captureThumbnailRequested() const = 0;
```
Comment (igorkorsukov) [inline src/project/ithumbnailcreator.h:18]: here are the details of the implementation revealed... I think it's not worth complicating it and solving it, I just suggest adding a comment and regrouping it somehow like this 

```
    virtual muse::Ret createThumbnail(const muse::io::path_t& path) = 0;

    // for internal use 
    virtual muse::async::Channel<muse::io::path_t> captureThumbnailRequested() const = 0;
    virtual void onThumbnailCreated(bool success) = 0;
```

File: src/project/projectmodule.cpp [hunk 8]
```diff
@@ -80,6 +82,7 @@ void ProjectModule::registerExports()
     ioc()->registerExport<IRecentFilesController>(moduleName(), m_recentFilesController);
     ioc()->registerExport<IOpenSaveProjectScenario>(moduleName(), new OpenSaveProjectScenario());
     ioc()->registerExport<IProjectFilesController>(moduleName(), m_actionsController);
+    ioc()->registerExport<IThumbnailCreator>(moduleName(), m_thumbnailCreator);
```
Comment (igorkorsukov) [inline src/project/projectmodule.cpp:85]: just for information, if the object is not used in other methods (for example, we don't need to do init in onInit), then we don't have to create a class member (this is not a problem))
```
ioc()->registerExport<IThumbnailCreator>(moduleName(), new ThumbnailCreator());
```

File: src/project/ithumbnailcreator.h (hunk 7 above)
Comment (embarc-gabriel) [reply src/project/ithumbnailcreator.h:20]: Done.

File: src/project/internal/thumbnailcreator.h (hunk 6 above)
Comment (embarc-gabriel) [reply src/project/internal/thumbnailcreator.h]: Sure! I don´t know why I forgot this. Thanks.

File: src/project/ithumbnailcreator.h (hunk 7 above)
Comment (igorkorsukov) [reply src/project/ithumbnailcreator.h:20]: mistake in grouping :)

Comment (embarc-gabriel) [reply src/project/ithumbnailcreator.h:19]: Done!

Comment (embarc-gabriel) [reply src/project/ithumbnailcreator.h:19]: Sorry for that.

**Review by igorkorsukov — APPROVED** [review]

**Pull request merged**